### PR TITLE
Complex Puzzle

### LINKED ISSUE #29: Compex Puzzle

### PR DESCRIPTION
**Pull request by ChaseBENNC9**
This will expand upon the interactions feature by allowing a scenario where the player needs to collect multiple items before they can actually use this. 
The most basic form is a door that requires two keys and a more complex idea is to be attaching objects together to reach something or to build a tool to help the player

## Progress
What i have so far is the player can pick up two different pieces and attach them to the wall only in the correct order. 
First the silver piece which is a piece of a valve. The piece should teleport in place
Then the wheel / handle can be attached.
Then the user can interact again and print to the screen that interaction was successful

### CODE REVIEW

**Comment by ChaseBENNC9** [comment]
The system for how the goal works when it has the correct items has been refactored.
I am now using the unity events which allows me to choose a method in the inspector that will be called.
this is a much cleaner process and means i dont need an endless if else check for each type of puzzle we have

**Comment by ChaseBENNC9** [comment]
- I have refactored alot of the code so the goal is split into two classes.  
Item goal as the parent and Complex goal as a child which has changed certain methods to fit the behaviour needed

File: AI Game Jam/Assets/Scripts/ItemGoal.cs [hunk 1]
```diff
@@ -1,38 +1,66 @@
...
+    protected UnityEvent goalAction; //the action will be called when the goal is interacted with and its requirements are met protected so it can be accessed by child classes
+
+    protected bool isComplex; //if the hint should be shown protected so it can be accessed by child classes
+    public bool IsComplex
+    {
+        get => isComplex;
+    } //Creates a get for the isComplex so other scripts can read its value
+
+
+
```
Comment (erokasyuart) [inline AI Game Jam/Assets/Scripts/ItemGoal.cs]: Too much whitespace

File: AI Game Jam/Assets/Scripts/ItemGoal.cs [hunk 2]
```diff
@@ -1,38 +1,66 @@
...
     public void ShowHint(bool show, string hintstring = "")
     {
         hint.GetComponent<TMP_Text>().text = hintstring;
         hint.SetActive(show);
     }
 
+    public void GoalActionDrainWater() //This method will be run by the goal action when the goal is completed if it is set in the inspector
+    {
+        Debug.Log("Goal Action Invoked");
+        Debug.Log("Wheel turned and water drained");
```
Comment (erokasyuart) [inline AI Game Jam/Assets/Scripts/ItemGoal.cs]: All the debug logs must go!!

File: AI Game Jam/Assets/Scripts/PlayerInteraction.cs [hunk 3]
```diff
@@ -49,14 +50,29 @@ void OnTriggerEnter(Collider other)
...
         {
-            if (item != null && item.GetComponent<Item>().Type ==  other.gameObject.transform.parent.GetComponent<ItemGoal>().itemName
+            if (
+                item != null
+                && item.GetComponent<Item>().Type.ToLower()
+                    == other.gameObject.transform.parent.GetComponent<ItemGoal>().itemName.ToLower()
             ) //if the item name is the same as the goal name
             {
-                other.gameObject.transform.parent.GetComponent<ItemGoal>().ShowHint(true, other.gameObject.transform.parent.GetComponent<ItemGoal>().hintWithItem); //show the hint for the goal wnen the player has the item
+                other
```
Comment (erokasyuart) [inline AI Game Jam/Assets/Scripts/PlayerInteraction.cs]: This is weird indentation...

**Review by erokasyuart — CHANGES_REQUESTED** [review]
I have no issue with the code other than the debug logs and general housekeeping.

**Review by erokasyuart — APPROVED** [review]

**Pull request merged**